README.md
# An admin panel for creating & managing sites in the openstad ecosystem

## Prerequisites
 - [Git](https://git-scm.com/)
 - [Node.js and npm](https://nodejs.org/en/)


#### 1. Set .env values
```
PORT=7777
PRODUCTION_SERVER_IP=132.123123 #For d
WILDCARD_HOST=cms.staging.amsterdam.nl
API_URL=
APP_URL=
USER_API=http://localhost:4000/api/admin
USER_API_CLIENT_ID=ccc
USER_API_CLIENT_SECRET=ccccc
COOKIE_SECURE_OFF=yes #notforlive
SESSION_SECRET= #random
SITE_ID=6 #site Id from the openstad api
SITE_API_KEY=xxxxxxx #site Id from the api
BASIC_AUTH_USER= #basic auth pass, whatever you prefer
BASIC_AUTH_PASSWORD=
EXTERNAL_SITE_REPO=API_URL/api/repo
FRONTEND_URL=http://cms.staging.amsterdam.nl
PUBLIC_IP=
```

#### 2. Run NPM install

```
npm i
```


#### 3. Run dev server

```
npm run dev
```


CHANGELOG.md
# Changelog

## v0.41.0
* Use networking.k8s.io/v1 version for ingress
  * IMPORTANT: This requires Kubernetes version v1.19 due to switching from networking.k8s.io/v1beta1 to networking.k8s.io/v1. 
  * See https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122 for more details.

## v0.40.0
Update version number

## v0.22.0
* Fix site delete error in k8s
* Move docker builds from travis to github actions
* Add email image header setting field to authentication tab (using filepond to upload image via the image server)
* Validate email address on special characters and reject the follow characters: ! and =

## v0.21.0
* Update react admin
* Update proxy call to API
* Add 'anonymize site' functionality
* Add extraData to newsletter export

## v0.20.0
* Rewrite of ingress logic and adding support for subdir sites (domain.com/site1 &&domain.com/site2) for Kubernetes environments
* For new sites set 2FA on by default for admin, editors & moderators

## v0.18.0
* Fix shared images between copies of sites

## v0.17.0
* Load react-admin from cdn - OPENSTAD_REACT_ADMIN_CDN in .env is now an optional overrule
* When creating a new site set contactEmail from the form

## v0.16.1 (2020-06-08)
* Add option to configure SMS sender and message.

## v0.16.0 (2020-05-25)
* Fix domain validation on site creation.
* Add phonenumber and local auth as experimental auth methods
* Turn on choice guide in react admin panel 

## v0.15.0 (2020-04-26)
* Add two factor configuration of text on screens and what role per oAuth client should validate with a Two Factor token
* Add ability to  reset two factor configuration in user edit form
* Allow for site names to have a - sign

## v0.14.1 (2021-04-19)
* Fix Site URL input

## v0.14.0 (2020-04-06)
* Cut the name for mongodb name shorter, in same cases it was too long causing an db creation error
* On input even in domain field strip special characters 
* Add options for selecting  for voting type: count per theme and budgetting per theme in site vote setting

## v0.13.0 (2021-03-17)
* Remove checkboxes in import and export for ideas and choice guides
* Update Openstad React Admin JS to v0.10

## v0.12.0 (2021-02-23)
* Make URL in table overview table a link to the url
* For localhost skip DNS test, allowing development environments to create sites
* React admin via jsdelivr

## v0.11.1 (2021-02-02)
* Remove publish button, not needed anymore happens automatically

## v0.11.0 (2021-01-27)
* Allowing to set ALLOW_HTTP_URLS to ON will allow http:// edit
* Allowing site with urls with subdirectories like domain.com/site1 domain.com/site2

## v0.10.3 (2020-12-16)
* Only allow to select an authentication method that uses e-mail if the user wants to make e-mail a required field.
* For default site don't allow editing of basic auth or delete.

## v0.10.2 (2020-12-09)
* Only allow to select an authentication method that uses e-mail if the user wants to make e-mail an required field.

## v0.10.1 (2020-12-09)
* Log errors while uploading images

## v0.10.0 (2020-12-09)
* Only allow alphanumeric characters in the uniqueidentifier otherwise will cause issues with copying sites

## v0.9.0 (2020-11-17)
* When creating a user, in the role editing table also display the site domain next to the client ID
* Bugfix: fix importing site by adding site name field

## v0.8.2 (2020-11-03)
* Update react-admin
* Update NPM modules for security
* Select for setting defaultRoleId in admin panel


## v0.8.2 (2020-11-03)
* Update react-admin
* Update NPM modules for security
* Select for setting defaultRoleId in admin panel

## v0.8.0-PRERLEASE (2020-17-07)
* Set content-length with Bufferlength for body api proxy, otherwise it breaks with special characters
* Add slash to formatting tmp directory in site export  
* Update corresponding ingress when site URL is changed
* Delete corresponding ingress when a site is deleted

## 0.7.5 (2020-09-29)
* Update openstad logo

## 0.7.4 (2020-09-29)
* Fix mistakes in explanations for idea thank you email

## 0.7.3 (2020-09-29)
* Update hostname in CMS config on url update

## 0.7.2 (2020-09-29)
* Hide old idea import & export buttons

## 0.7.1 (2020-09-16)
* Fix kubernetes site creation, by skipping broken DNS check

## 0.7.0 (2020-09


public/js/main.js
'use strict';


initHideFlash();


function initHideFlash() {
  $('.flash-container .close-button').click(function() {
    $(this).closest('.flash-container').remove();
  });

  setTimeout(function() {
    $('.flash-container').remove();
  }, 5000);
}

/*
function initExpand () {
  $( ".expand" ).each(function( index ) {
  });
}

 */

function displayDomainFields () {
  var domainType = $('input[name="domain-type"]:checked').val();
  $('.display-for-subdir, .display-for-domain, .display-for-subdomain').hide();

  // console.log('.display-for'+domainType);
  $('.display-for-'+domainType).show();
}

$(function () {


  displayDomainFields ()

  // # CREATE SITE FORM VALIDATION
  var existingDomains = [];

  $.validator.addMethod("uniqueDomain", function (value, element) {
    var isValid = false;

    var domainType = $('input[name="domain-type"]:checked').val();
    var wildcardHost = $('input[name="wildcardHost"]').val();
    var domainValue = domainType === 'subdomain' ? value + wildcardHost : value;
    var existingDomains = $('input[name="domain"]').data('existing-domains').split(',');

    return existingDomains.indexOf(domainValue.replace('https://', '').replace('http://', '').replace('www.', '')) === -1;
  });

  $('.domain-type').change(function (event) {
    displayDomainFields();
  });

  $('#create-site-form').validate({
    rules: {
      domain: {
        uniqueDomain: true,
        required: true
      }
    },
    messages: {
      domain: {
        uniqueDomain: 'Domain already exists'
      }
    }
  });

  // filter everythings but dots and alphanumeric from domain fields, also no other specialchars allowed
  $(".valid-domain-character").keypress(function(event) {
      var key = event.which;
      var keychar = String.fromCharCode(key).toLowerCase();
      var allowedCharacters = "abcdefghijklmnopqrstuvwxyz0123456789-.:";

      if ($(this).hasClass('valid-domain-character-allow-slash')) {
        allowedCharacters = allowedCharacters + '/';
      }

      if ((allowedCharacters).indexOf(keychar) === -1) {
         event.preventDefault();
         return false;
      }
  });




  $(".valid-domain-character").on('input', function(event) {
    //also enfore lowercase
    var lowercaseValue = $(this).val().toLowerCase();
    var regex = /[^a-zA-Z0-9-.:_]/g;


    if ($(this).hasClass('valid-domain-character-allow-slash')) {
      regex = /[^a-zA-Z0-9-.//:_]/g;
    }

    // remove all chars that are not alpha numeric
    lowercaseValue = lowercaseValue.replace(regex, "");

    $(this).val(lowercaseValue);
  });

  // #3. FORM VALIDATION

  $('.validate-form').each(function() {
    $(this).validate();
  })

  $('#dataTable1').dataTable();

  // #4. DATE RANGE PICKER

  //$('input.single-daterange').daterangepicker({ "singleDatePicker": true });
  //$('input.multi-daterange').daterangepicker({ "startDate": "03/28/2017", "endDate": "04/06/2017" });

  // #5. DATATABLES

  if ($('#formValidate').length) {
    $('#formValidate').validator();
  }

  function userColumns () {
    return  [
       { data: "id" },
       { data: "firstName" },
       { data: "lastName" },
       { data: "email" },
       { data: "id",
         render: function (val) {
           return '<a href="/admin/user/'+val+'" target="_blank">Edit</a>'
        }
      }
    ];
  }

  function uniqueCodeColumns () {
    return  [
       { data: "id" },
       { data: "code" },
       { data: "userId", render:  function (val) { return val ? 'yes' : 'no'  } },
       { data: function(row, type, set) {
          return row;
       }, render:  function (val, row) {
         return val.userId ? '<a href="/admin/unique-code/reset/'+val.id+'">Reset </a>' : '' }
       },
    ];
  }

  if ($('#dataTable-ajax').length) {
    var dataColumns = $('#dataTable-ajax').attr('data-custom-columns');

    $('#dataTable-ajax').DataTable({
      buttons: ['copy', 'excel', 'pdf'],
      processing: true,
      paging: true,
      ordering: false,
      iDisplayLength: 50,
      pageLength: 50,
      ajax: function ( data, callback, settings ) {
          var apiUrl =  $('#dataTable-ajax').attr('data-src');

          var params = {
            offset: data.start,
            limit: data.length,
            //order: 'id,'
          };

          if (data.search && data.search.value && data.search.value.length >0) {
            params.search = data.search.value;
          }

          var apiQuery = jQuery.param(params);
          apiUrl = apiUrl +'?a=1&'+ apiQuery;

          $.ajax({
              url:apiUrl,
              // dataType: 'text',
              type: 'get',
              contentType: 'JSON',
              success: function( responseData, textStatus, jQxhr ){
                  callback({
                      // draw: data.draw,
                      data: responseData.data,
                      recordsTotal:  responseData.total,//data.length,
                      recordsFiltered:  responseData.total,// data.length
                  });
              },
              error: function( jqXhr, textStatus, errorThrown ){
              }
          });
      },
      serverSide: true,
    /*  columns: [
        { data: "id" },
        { data: "code" },
        { data: "userId", render:  function (val) { return val ? 'yes' : 'no'  } },
      ]*/
      columns: eval(dataColumns)()
    });

}


  // #7. FORM STEPS FUNCTIONALITY

  $('.step-trigger-btn').on('click', function () {
    var btn_href = $(this).attr('href');
    $('.step-trigger[href="' + btn_href + '"]').click();
    return false;
  });

  // FORM STEP CLICK
  $('.step-trigger').on('click', function () {
    var prev_trigger = $(this).prev('.step-trigger');
    if (prev_trigger.length && !prev_trigger.hasClass('active') && !prev_trigger.hasClass('complete')) return false;
    var content_id = $(this).attr('href');
    $(this).closest('.step-triggers').find('.step-trigger').removeClass('active');
    $(this).prev('.step-trigger').addClass('complete');
    $(this).addClass('active');
    $('.step-content').removeClass('active');
    $('.step-content' + content_id).addClass('active');
    return false;
  });


  if ($('.select2').length) {
    $('.select2').select2();
  }

  // Tasks foldable trigger
  $('.tasks-header-toggler').on('click', function () {
    $(this).closest('.tasks-section').find('.tasks-list-w').slideToggle(100);
    return false;
  });

  // #21. Onboarding Screens Modal
  $('.onboarding-modal.show-on-load').modal('show');
  if ($('.onboarding-modal .onboarding-slider-w').length) {
    $('.onboarding-modal .onboarding-slider-w').slick({
      dots: true,
      infinite: false,
      adaptiveHeight: true,
      slidesToShow: 1,
      slidesToScroll: 1
    });

    $('.onboarding-modal').on('shown.bs.modal', function (e) {
      $('.onboarding-modal .onboarding-slider-w').slick('setPosition');
    });
  }


  // Init filepond
  var filepondFieldset = document.querySelector('.filepondFieldset');
  if (filepondFieldset) {
    FilePond.registerPlugin(FilePondPluginImagePreview);
    FilePond.registerPlugin(FilePondPluginFilePoster);
    FilePond.registerPlugin(FilePondPluginImageValidateSize);
    FilePond.registerPlugin(FilePondPluginFileValidateType);
    FilePond.registerPlugin(FilePondPluginImageExifOrientation);

    var image = filepondFieldset.dataset.image;
    var uploadedFiles = [];
    if (image) {
      uploadedFiles.push({
        source: '{"url":"'+ image +'"}',
        options : {
          type: 'local',
          file: {
            name: image,
          },
          metadata: {
            poster: image,
          }
        }
      });
    }

    var filePondSettings = {
      // set allowed file types with mime types
      acceptedFileTypes: ['image/*'],
      allowFileSizeValidation: true,
      maxFileSize: '8mb',
      name: 'image',
      maxFiles: 1,
      allowBrowse: true,
      files: uploadedFiles,
      filePosterMaxHeight: 256,
      imagePreviewMaxHeight: 256,
      server: {
        process: '/image',
        revert: null
      },
      labelIdle: "Sleep afbeelding naar deze plek of <span class='filepond--label-action'>klik hier</span>",
      labelInvalidField: "Field contains invalid files",
      labelFileWaitingForSize: "Wachtend op grootte",
      labelFileSizeNotAvailable: "Grootte niet beschikbaar",
      labelFileCountSingular: "Bestand in lijst",
      labelFileCountPlural: "Bestanden in lijst",
      labelFileLoading: "Laden",
      labelFileAdded: "Toegevoegd", // assistive only
      labelFileLoadError: "Fout bij het uploaden",
      labelFileRemoved: "Verwijderd", // assistive only
      labelFileRemoveError: "Fout bij het verwijderen",
      labelFileProcessing: "Laden",
      labelFileProcessingComplete: "Afbeelding geladen",
      labelFileProcessingAborted: "Upload cancelled",
      labelFileProcessingError: "Error during upload",
      labelFileProcessingRevertError: "Error during revert",
      labelTapToCancel: "tap to cancel",
      labelTapToRetry: "tap to retry",
      labelTapToUndo: "tap to undo",
      labelButtonRemoveItem: "Verwijderen",
      labelButtonAbortItemLoad: "Abort",
      labelButtonRetryItemLoad: "Retry",
      labelButtonAbortItemProcessing: "Verwijder",
      labelButtonUndoItemProcessing: "Undo",
      labelButtonRetryItemProcessing: "Retry",
      labelButtonProcessItem: "Upload"
    }

    FilePond.create(filepondFieldset, filePondSettings);
  }
});
